### PR TITLE
[F] Audio Transcription

### PR DESCRIPTION
**Pull request by mmatongo**

### CODE REVIEW

File: internal/speech/speech.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	op, err := client.LongRunningRecognize(ctx, req)
+	if err != nil {
+		return "", fmt.Errorf("failed to start long running recognition: %v", err)
+	}
+
+	resp, err := op.Wait(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to get long running recognition results: %v", err)
+	}
```
Comment (ccoVeille) [inline internal/speech/speech.go]:
```suggestion
	client, err := speech.NewClient(ctx, clientOpts...)
	if err != nil {
		return "", fmt.Errorf("failed to create client: %w", err)
	}
	defer client.Close()

	audioInfo, err := getAudioInfo(filename)
	if err != nil {
		return "", fmt.Errorf("failed to process audio file: %e", err)
	}

	gcsURI, err := uploadToGCS(ctx, opts.Bucket, filename)
	if err != nil {
		return "", fmt.Errorf("failed to upload to GCS: %e", err)
	}

	diarizationConfig := &speechpb.SpeakerDiarizationConfig{
		EnableSpeakerDiarization: opts.EnableDiarization,
		MinSpeakerCount:          int32(opts.MinSpeakers),
		MaxSpeakerCount:          int32(opts.MaxSpeakers),
	}

	req := &speechpb.LongRunningRecognizeRequest{
		Config: &speechpb.RecognitionConfig{
			Encoding:                   getEncoding(audioInfo.format),
			SampleRateHertz:            int32(audioInfo.sampleRate),
			AudioChannelCount:          int32(audioInfo.numChannels),
			LanguageCode:               opts.LanguageCode,
			EnableAutomaticPunctuation: true,
			UseEnhanced:                true,
			EnableWordConfidence:       true,
			Model:                      "latest_long",
			DiarizationConfig:          diarizationConfig,
		},
		Audio: &speechpb.RecognitionAudio{
			AudioSource: &speechpb.RecognitionAudio_Uri{
				Uri: gcsURI,
			},
		},
	}

	op, err := client.LongRunningRecognize(ctx, req)
	if err != nil {
		return "", fmt.Errorf("failed to start long running recognition: %w", err)
	}

	resp, err := op.Wait(ctx)
	if err != nil {
		return "", fmt.Errorf("failed to get long running recognition results: %w", err)
	}
```

File: internal/speech/processors.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open WAV file: %v", err)
+	}
+	defer file.Close()
+
+	decoder := wav.NewDecoder(file)
+	if !decoder.IsValidFile() {
+		return nil, fmt.Errorf("invalid WAV file")
+	}
```
Comment (ccoVeille) [inline internal/speech/processors.go]:
```suggestion
func processMp3(filename string) (*audioInfo, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("failed to open MP3 file: %w", err)
	}

	defer file.Close()

	decoder, err := mp3.NewDecoder(file)
	if err != nil {
		return nil, fmt.Errorf("failed to create MP3 decoder: %w", err)
	}

	return &audioInfo{
		sampleRate:  decoder.SampleRate(),
		numChannels: 2,
		bitDepth:    0, // http://blog.bjrn.se/2008/10/lets-build-mp3-decoder.html
		format:      "MP3",
	}, nil
}

func processFlac(filename string) (*audioInfo, error) {
	file, err := flac.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("failed to open FLAC file: %w", err)
	}
	defer file.Close()

	return &audioInfo{
		sampleRate:  int(file.Info.SampleRate),
		numChannels: int(file.Info.NChannels),
		bitDepth:    int(file.Info.BitsPerSample),
		format:      "FLAC",
	}, nil
}

func processWav(filename string) (*audioInfo, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("failed to open WAV file: %w", err)
	}
	defer file.Close()

	decoder := wav.NewDecoder(file)
	if !decoder.IsValidFile() {
		return nil, errors.New("invalid WAV file")
	}
```

File: internal/speech/processors.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+type audioInfo struct {
+	sampleRate  int
+	numChannels int
+	bitDepth    int
+	format      string
+}
+
+func getAudioInfo(filename string) (*audioInfo, error) {
+	ext := strings.ToLower((filepath.Ext(filename)))
```
Comment (ccoVeille) [inline internal/speech/processors.go]:
```suggestion
	ext := strings.ToLower(filepath.Ext(filename))
```

File: internal/speech/google.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	w := client.Bucket(bucket).Object(objectName).NewWriter(ctx)
+	if _, err = io.Copy(w, f); err != nil {
+		return "", fmt.Errorf("failed to copy file to GCS: %v", err)
+	}
+	if err := w.Close(); err != nil {
+		return "", fmt.Errorf("failed to close GCS writer: %v", err)
+	}
+
+	return fmt.Sprintf("gs://%s/%s", bucket, objectName), nil
+}
```
Comment (ccoVeille) [inline internal/speech/google.go]:
```suggestion
func uploadToGCS(ctx context.Context, bucket, filename string) (string, error) {
	client, err := storage.NewClient(ctx)
	if err != nil {
		return "", fmt.Errorf("failed to create storage client: %w", err)
	}
	defer client.Close()

	f, err := os.Open(filename)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %w", err)
	}
	defer f.Close()

	objectName := filepath.Base(filename)
	w := client.Bucket(bucket).Object(objectName).NewWriter(ctx)
	if _, err = io.Copy(w, f); err != nil {
		return "", fmt.Errorf("failed to copy file to GCS: %w", err)
	}
	if err := w.Close(); err != nil {
		return "", fmt.Errorf("failed to close GCS writer: %w", err)
	}

	return fmt.Sprintf("gs://%s/%s", bucket, objectName), nil
}
```

File: README.md [hunk 5]
```diff
@@ -68,7 +68,7 @@ Content: More information...
 
 ## Features
 
-- Supports multiple content types: HTML, PDF, CSV, JSON, YAML, DOCX, and Markdown
+- Supports multiple content types: HTML, PDF, CSV, JSON, YAML, DOCX, PPTX, Markdown, Plaintext, MP3, FLAC and WAVE
```
Comment (ccoVeille) [inline README.md]:
```suggestion
- Supports multiple content types: HTML, PDF, CSV, JSON, YAML, DOCX, PPTX, Markdown, Plaintext, MP3, FLAC, and WAVE
```

**Comment by ccoVeille** [comment]
A few suggestions

**Comment by mmatongo** [comment]
> A few suggestions

Thanks @ccoVeille, I've added the changes suggested.

**Comment by ccoVeille** [comment]
You are welcome

Unless you plan to squash your commits, I would like to report there is a typo in your last commit message

it's transcribing not trNAscribing

**Comment by mmatongo** [comment]
> You are welcome
> 
> Unless you plan to squash your commits, I would like to report there is a typo in your last commit message
> 
> it's transcribing not trNAscribing

thanks, I'll fix that

**Review by mmatongo — COMMENTED** [review]
Feature development is complete.

**Pull request merged**